File: tests/list.ts

#include "list.h"

#include <check.h>
#include <stddef.h>

#test empty
	List* list = listCreate();
	ck_assert_ptr_null(listIterate(list, NULL));
	listFree(&list);
	ck_assert_ptr_null(list);

#test append
	List* list = listCreate();

	int* item;
	item = listAppend(list, sizeof(int));
	*item = 123;
	item = listAppend(list, sizeof(int));
	*item = 456;
	item = listAppend(list, sizeof(int));
	*item = 789;

	item = listIterate(list, NULL);
	ck_assert_int_eq(*item, 123);
	item = listIterate(list, item);
	ck_assert_int_eq(*item, 456);
	item = listIterate(list, item);
	ck_assert_int_eq(*item, 789);

	listFree(&list);
	ck_assert_ptr_null(list);

#test insert
	List* list = listCreate();

	int *item, *item2;
	item = listAppend(list, sizeof(int));
	*item = 123;
	item2 = listAppend(list, sizeof(int));
	*item2 = 789;
	item = listInsert(list, item, sizeof(int));
	*item = 456;

	item = listIterate(list, NULL);
	ck_assert_int_eq(*item, 123);
	item = listIterate(list, item);
	ck_assert_int_eq(*item, 456);
	item = listIterate(list, item);
	ck_assert_int_eq(*item, 789);

	listFree(&list);
	ck_assert_ptr_null(list);

File: src/list.c
#include "list.h"

#include <stdlib.h>

typedef struct Element
{
	struct Element* next;
	char data[];
} Element;

List* listCreate()
{
	Element* root = malloc(sizeof(Element));
	root->next = NULL;
	return (List*)root;
}

void* listAppend(List* list, size_t elementSize)
{
	void* element;
	void* nextElement;

	for (element = ((Element*)list)->data; (nextElement = listIterate(list, element)); element = nextElement)
	{
	}
	
	return listInsert(list, element, elementSize);
}

void* listInsert(List* list, const void* afterElement, size_t elementSize)
{
	Element** newElementPtr = &(((Element*)afterElement) - 1)->next;
	Element* nextElement = *newElementPtr;
	(*newElementPtr) = malloc(sizeof(Element) + elementSize);
	(*newElementPtr)->next = nextElement;
	return &(*newElementPtr)->data;
}

void* listIterate(List* list, const void* element)
{
	Element* next;

	if (element)
	{
		next = (((Element*)element) - 1)->next;
	}
	else
	{
		next = ((Element*)list)->next;
	}

	if (next)
	{
		return &next->data;
	}
	else
	{
		return NULL;
	}
}

void listFree(List** list)
{
	Element* element = (Element*)(*list);

	while (element)
	{
		Element* next = element->next;
		free(element);
		element = next;
	}

	*list = NULL;
}
